cmds/free/free.go

```diff
@@ -0,0 +1,343 @@
+// Copyright 2012-2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// free displays information about total, used and available physical memory and
+// swap space.
+//
+// Synopsis:
+//     free [-k] [-m] [-g] [-t] [-h] [-json]
+//
+// Description:
+//     Read memory information from /proc/meminfo and display a summary for
+//     physical memory and swap space. The unit options use powers of 1024.
+//
+// Options:
+//     -k: display the values in kibibytes
+//     -m: display the values in mebibytes
+//     -g: display the values in gibibytes
+//     -t: display the values in tebibytes
+//     -h: display the values in human-readable form
+//     -json: use JSON output
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strconv"
+)
+
+const meminfoFile = "/proc/meminfo"
+
+var humanOutput = flag.Bool("h", false, "Human output: show automatically the shortest three-digits unit")
+var inBytes = flag.Bool("b", false, "Express the values in bytes")
+var inKB = flag.Bool("k", false, "Express the values in kibibytes (default)")
+var inMB = flag.Bool("m", false, "Express the values in mebibytes")
+var inGB = flag.Bool("g", false, "Express the values in gibibytes")
+var inTB = flag.Bool("t", false, "Express the values in tebibytes")
+var toJSON = flag.Bool("json", false, "Use JSON for output")
+
+type unit uint
+
+const (
+	// B is bytes
+	B unit = 0
+	// KB is kibibytes
+	KB = 10
+	// MB is mebibytes
+	MB = 20
+	// GB is gibibytes
+	GB = 30
+	// TB is tebibytes
+	TB = 40
+)
+
+var units = [...]string{"B", "K", "M", "G", "T"}
+
+// FreeConfig is a structure used to configure the behaviour of Free()
+type FreeConfig struct {
+	Unit        unit
+	HumanOutput bool
+	ToJSON      bool
+}
+
+// the following types are used for JSON serialization
+type mainMemInfo struct {
+	Total     uint64 `json:"total"`
+	Used      uint64 `json:"used"`
+	Free      uint64 `json:"free"`
+	Shared    uint64 `json:"shared"`
+	Cached    uint64 `json:"cached"`
+	Buffers   uint64 `json:"buffers"`
+	Available uint64 `json:"available"`
+}
+
+type swapInfo struct {
+	Total uint64 `json:"total"`
+	Used  uint64 `json:"used"`
+	Free  uint64 `json:"free"`
+}
+
+// MemInfo represents the main memory and swap space informatio in a structured
+// manner, suitable for JSON encoding.
+type MemInfo struct {
+	Mem  mainMemInfo `json:"mem"`
+	Swap swapInfo    `json:"swap"`
+}
+
+type meminfomap map[string]uint64
+
+// meminfo returns a mapping that represents the fields contained in
+// /proc/meminfo
+func meminfo() (meminfomap, error) {
```

Comment:
`meminfo` and `meminfoFromBytes` should be used to "free_linux.go" because they a specific to Linux.

Reply:
Fair point. Are we considering linux- and non-linux-specific code for u-root?

Reply:
As far as I know, at the moment no one is using non-linux builds of u-root. But, we a trying to keep the code separated enough so it is easy to port to other archs and OSes in the future.

Reply:
makes sense, I'm moving it to its own file